packages/@lwc/shared/src/fields.ts

```diff
@@ -4,38 +4,35 @@
  * SPDX-License-Identifier: MIT
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
  */
-import { create, isUndefined } from './language';
+import { isUndefined } from './language';
 import { hasNativeSymbolSupport } from './symbol';
 
-/*
- * The __type property on HiddenField doesn't actually exists at runtime. It is only used to store
- * the type of the associated field value.
- */
-export type HiddenField<T> = { __type: T };
-
-// This method abstracts the creation of symbols, so we can fallback to strings when native symbols
-// are not supported.
-export function createHiddenField<T = unknown>(key: string, namespace: string): HiddenField<T> {
-    return (hasNativeSymbolSupport ? Symbol(key) : `$$lwc-${namespace}-${key}$$`) as any;
-}
+export type HiddenField<T> = {
+    set: (o: any, value: T) => void;
+    get: (o: any) => T | undefined;
+};
 
-const hiddenFieldsMap: WeakMap<any, Record<any, any>> = new WeakMap();
+// Typically there are lots of objects we're assigning hidden fields on, but few unique fields.
+// So we create fewer objects and do less work to map from Field -> Object -> Value rather than
+// from Object -> Field -> Value
+const hiddenFieldMaps: Map<string | symbol, WeakMap<any, any>> = new Map();
```

Comment:
First of all, this is a great performance improvement. 

A couple of observations:
- You can directly return the `valuesByObject` instead of creating a brand new object. This is because `WeakMap<any, T>` is a superset of `HiddenField<T>`. 
- I am now questioning the actual value of `createHiddenField` over using a plain `WeakMap`. I don't see where `createHiddenField` reuse the same key or namespace throughout the codebase. In other words `WeakMap` isn't reused at all in the LWC engine.
 
I don't see any reason why we can't completely drop `createHiddenField` and replace it with inline `WeakMap` and `WeakSet`.

Reply:
> You can directly return the `valuesByObject` instead of creating a brand new object. This is because `WeakMap<any, T>` is a superset of `HiddenField<T>`. 

🤦  I didn't think of this!

But yes, it feels like we could drop `createHiddenField` and just have `createSymbol` or something, to avoid duplicating the code that chooses strings vs symbols. I'll try some refactoring.